src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java

```diff
@@ -0,0 +1,302 @@
+      if (!s.substring(0, 1).toLowerCase().equals(s.substring(0, 1))) {
+        HaxeSemanticError.addError(element, new HaxeSemanticError("Package name '" + s + "' must start with a lower case character"));
+      }
+    }
+
+    if (!packageName.equals(actualPackage)) {
+      HaxeSemanticError.addError(
+        element,
+        new HaxeSemanticError(
+          "Invalid package name! '" + packageName + "' should be '" + actualPackage + "'",
+          new HaxeSemanticIntentionAction("Fix package") {
+            @Override
+            public void run() {
+              Document document =
+                PsiDocumentManager.getInstance(element.getProject()).getDocument(element.getContainingFile());
+
+              if (expression != null) {
+                TextRange range = expression.getTextRange();
+                document.replaceString(range.getStartOffset(), range.getEndOffset(), actualPackage);
+              }
+              else {
+                int offset =
+                  element.getNode().findChildByType(HaxeTokenTypes.OSEMI).getTextRange().getStartOffset();
+                document.replaceString(offset, offset, actualPackage);
+              }
+            }
+          }
+        )
+      );
+    }
+  }
+}
+
+class PsiFileUtils {
+
+  static public String getListPath(List<PsiFileSystemItem> range) {
+    String out = "";
+    for (PsiFileSystemItem item : range) {
+      if (out.length() != 0) out += "/";
+      out += item.getName();
+    }
+    return out;
+  }
+
+  static public List<PsiFileSystemItem> getRange(PsiFileSystemItem from, PsiFileSystemItem to) {
+    LinkedList<PsiFileSystemItem> items = new LinkedList<PsiFileSystemItem>();
+    PsiFileSystemItem current = to;
+    while (current != null) {
+      items.addFirst(current);
+      if (current == from) break;
+      current = current.getParent();
+    }
+    return items;
+  }
+
+  static public PsiFileSystemItem findRoot(PsiFileSystemItem file) {
+    //HaxelibModuleManager
+    Project project = file.getProject();
+    Module[] modules = ModuleManager.getInstance(project).getModules();
+    List<VirtualFile> roots = new ArrayList<VirtualFile>();
+    for (Module module : modules) {
+      Sdk sdk = ModuleRootManager.getInstance(module).getSdk();
+      roots.addAll(Arrays.asList(sdk.getRootProvider().getFiles(OrderRootType.CLASSES)));
+    }
+    roots.addAll(Arrays.asList(ProjectRootManager.getInstance(project).getContentSourceRoots()));
+
+    PsiFileSystemItem current = file;
+    while (current != null) {
+      for (VirtualFile root : roots) {
+        //System.out.println(root.getCanonicalPath());
+        //System.out.println(current.getVirtualFile().getCanonicalPath());
+        if (root.getCanonicalPath().equals(current.getVirtualFile().getCanonicalPath())) return current;
+      }
+      current = current.getParent();
+    }
+    return null;
+  }
+
+  static public String getDirectoryPath(PsiDirectory dir) {
+    String out = "";
```

Comment:
I'm iterating from the leaf to the root iterating over all the source roots to get the nearest source root in the ancestors path. And it would get to the disk root in the case its not founding any source root. Maybe in the case the source root is not found I could return null or something.
But if there is an available root it should work. At least for physical files. Suggestions to improve this are welcomed.